### PR TITLE
Add topic support

### LINKED ISSUE #219: Repository topic/labeling globally and within an organisation
When the number of repositories of an organisation grows to an extent, it's going to be difficult to keep the repos well-organised and easy to find. In directory-oriented VCS like Subversion, that can be intuitively handled by directory hierarchy. Something like labels for Git repository would help improving the user experience.

Ideas?

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/39507145-repository-labeling-within-an-organisation?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F47456670&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by lunny**
This PR will view a repository's topics and let repositorys owners to manage topics. 

Will fix #219 

image

image

image

image

### CODE REVIEW

**Comment by lunny** [comment]
I think it's ready for review.

**Comment by jonasfranz** [comment]
@lunny conflicted

**Comment by jonasfranz** [comment]
IMHO the color of the badges do not really fit to the design of gitea. We should use a sort of blue color (like links) or a green color like the logo.

**Comment by lunny** [comment]
conflict resolved. I will try to find another tag color.

**Comment by lunny** [comment]
@JonasFranzDEV changed to green basic style.

image

image

**Comment by jonasfranz** [comment]
When I create a new repo and add some topics it looks like:
screenshot-2018-3-29 as

When I reload it looks like: 
screenshot-2018-3-29 as 1

File: public/js/index.js [hunk 1]
```diff
@@ -2122,3 +2123,80 @@ function initNavbarContentToggle() {
...
+        },
+        className: {
+            label: 'ui green basic label'
+        },
+        apiSettings: {
+            url: suburl + '/api/v1/topics/search?q={query}',
+            throttle: 500,
+            cache: false,
+            onResponse: function(res) {
+                console.log(res)
```
Comment (appleboy) [inline public/js/index.js]: remove debug message?

File: public/js/index.js [hunk 2]
```diff
@@ -2122,3 +2123,80 @@ function initNavbarContentToggle() {
...
+                        formattedResponse.results.push({"description": res.topics[i].Name, "data-value":res.topics[i].Name})
+                    }
+                }
+
+                return formattedResponse;
+            },
+        },
+    });
+
+    $("span.sizer").change(function(){
```
Comment (appleboy) [inline public/js/index.js]: remove debug message.

**Review by appleboy — CHANGES_REQUESTED** [review]

**Review by appleboy — APPROVED** [review]

File: models/repo.go [hunk 3]
```diff
@@ -199,6 +199,7 @@ type Repository struct {
 	Size          int64              `xorm:"NOT NULL DEFAULT 0"`
 	IndexerStatus *RepoIndexerStatus `xorm:"-"`
 	IsFsckEnabled bool               `xorm:"NOT NULL DEFAULT true"`
+	Topics        []string           `xorm:"TEXT JSON"`
```
Comment (lafriks) [inline models/repo.go:202]: Why is this saved to database?

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/repo.go (hunk 3 above)
Comment (lunny) [reply models/repo.go:202]: For better performance when we list all topics for every repository on repositories list UI.

**Review by lafriks — APPROVED** [review]
Don't like duplicated staff in database as it usually brings it to sometimes not being updated but ok

**Comment by marbetschar** [comment]
@lafriks any chance of a new stable version any time soon which contains the topics? can't get it soon enough :heart_eyes:  :tada:  :rofl:

**Comment by lafriks** [comment]
@marbetschar 1.5.0 is planned to be release at the end of May but most probably that will be rc1 + 2-3 weeks to get to 1.5.0 so it's June :)

**Comment by marbetschar** [comment]
😱😱😱😱

🤣🤣

**Comment by Cyber1000** [comment]
Sorry to ask here, but I don't know where else:
Can you search after topics, or is it at least planned?
For now it seems I only can put topics on a repo and see it in the repo list, but I can't click on time like on github and get all repos relatet to this topic ...

Thanks!

**Comment by William-LP** [comment]
Hey, just tried under the last version 1.5.0-rc2 and it doesn't work :

image
---
image

Also the topics aren't clickable so this isn't possible to sort out the repo based on the topic

**Comment by lafriks** [comment]
Yes, searching and clicking on topics is currently not possible

**Comment by lunny** [comment]
This PR is a basic PR for topics. More PRs about topics are welcome!

**Comment by appleboy** [comment]
@Cyber1000 @William-LP See the PR https://github.com/go-gitea/gitea/pull/4505

File: templates/repo/home.tmpl [hunk 4]
```diff
@@ -5,7 +5,7 @@
 		{{template "base/alert" .}}
 		<div class="ui repo-description">
 			<div id="repo-desc">
-				{{if .Repository.DescriptionHTML}}<span class="description has-emoji">{{.Repository.DescriptionHTML}}</span>{{else if .IsRepositoryAdmin}}<span class="no-description text-italic">{{.i18n.Tr "repo.no_desc"}}</span>{{end}}
+				{{if .Repository.DescriptionHTML}}<span class="description has-emoji">{{.Repository.DescriptionHTML}}</span>{{else}}<span class="no-description text-italic">{{.i18n.Tr "repo.no_desc"}}</span>{{end}}
```
Comment (gabrielsimoes) [inline templates/repo/home.tmpl:8]: @lunny You accidentally removed #2167 here.

Comment (gabrielsimoes) [reply templates/repo/home.tmpl:8]: I addressed this on #5836

Comment (lunny) [reply templates/repo/home.tmpl:8]: Why we need ` if .IsRepositoryAdmin`?

Comment (gabrielsimoes) [reply templates/repo/home.tmpl:8]: The idea was to not display the "No Description" message to non-admins. That is, only the repo admin should see it.

**Pull request merged**